### PR TITLE
Fixed DelegateProxy to check if it's running on main thread

### PR DESCRIPTION
**Pull request by ryu1006**
Addressing issue: https://github.com/ReactiveX/RxSwift/issues/1874

### CODE REVIEW

File: RxSwift/Schedulers/MainScheduler.swift [hunk 1]
```diff
@@ -44,6 +45,13 @@ public final class MainScheduler : SerialDispatchQueueScheduler {
         }
     }
 
+    /// In case this method is running on a background thread it will throw an exception.
+    public class func ensureRunningOnMainThread(errorMessage: String? = nil) {
+        if !Thread.isMainThread {
```
Comment (freak4pc) [inline RxSwift/Schedulers/MainScheduler.swift]: I think a guard is nicer in this case
```suggestion
        guard Thread.isMainThread else { rxFatalError ... }
```

File: Platform/DispatchQueue+Extensions.swift [hunk 2]
```diff
@@ -7,15 +7,10 @@
...
 extension DispatchQueue {
-    private static var token: DispatchSpecificKey<()> = {
-        let key = DispatchSpecificKey<()>()
-        DispatchQueue.main.setSpecific(key: key, value: ())
-        return key
-    }()
-
     static var isMain: Bool {
-        return DispatchQueue.getSpecific(key: token) != nil
+        return Thread.isMainThread
```
Comment (freak4pc) [inline Platform/DispatchQueue+Extensions.swift]: I believe this entire file should be reverted to develop?

File: RxSwift/Schedulers/MainScheduler.swift (hunk 1 above)
Comment (ryu1006) [reply RxSwift/Schedulers/MainScheduler.swift]: Fixed

File: Platform/DispatchQueue+Extensions.swift (hunk 2 above)
Comment (ryu1006) [reply Platform/DispatchQueue+Extensions.swift]: Fixed

**Comment by kzaher** [comment]
Hi @ryu1006 ,

could you please take a look at Linux unit tests. I would like to release this with a patch release this weekend.

Otherwise looks ok to me.

@freak4pc Could you please smoke test this in a real app. I don't currently have access to any bigger codebase.

**Comment by freak4pc** [comment]
@kzaher I'll give it a whirl and let you know.

**Comment by ryu1006** [comment]
@kzaher checked and saw that isMainThread is not implemented in Linux's Foundation. Fortunately enough, majority of the instances for `ensureRunningOnMainThread` is only on files that ensure it's not running on Linux or is in Debug. Only `VirtualTimeScheduler` does not have these checks, so reverted it to the previous `ensureExecutingOnScheduler` function.

**Comment by freak4pc** [comment]
I've run ~500 unit tests of a larger codebase and everything seems to run smoothly. 
I guess the main pieces we have counting on DelegateProxy is RxDataSources - and that seems to work and pass tests with no issues.

File: Platform/DispatchQueue+Extensions.swift (hunk 2 above)
Comment (douknow) [reply Platform/DispatchQueue+Extensions.swift]: @freak4pc Just confused, can you explain why check by token is better than `Thread.isMainThread`?

**Pull request merged**